File: Sort/986. Interval List Intersections/986. Interval List Intersections.cpp
class Solution {
    static bool cmp(pair<int, int>&a, pair<int, int>&b) {
        if (a.first == b.first) {
            return a.second > b.second;
        } else {
            return a.first < b.first;
        }
    }
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& A, vector<vector<int>>& B) {
        vector<pair<int, int>>p;
        for (auto x : A) {
            p.push_back({x[0], 1});
            p.push_back({x[1], -1});
        }
        for (auto x : B) {
            p.push_back({x[0], 1});
            p.push_back({x[1], -1});
        }
        sort(p.begin(), p.end(), cmp);
        
        vector<vector<int>>rets;
        int start, end;
        int count = 0;
        for (auto x : p) {
            count += x.second;
            if (x.second == 1 && count == 2) {
                start = x.first;
            } else if (x.second == -1 && count == 1) {
                end = x.first;
                rets.push_back({start, end});
            }
        }
        return rets;
    }
};

File: SweepLine/986. Interval List Intersections/986. Interval List Intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
        // sweep line
        List<int[]> sweep = new ArrayList<>();
        for (int[] f : firstList) {
            sweep.add(new int[] {f[0], 1});
            sweep.add(new int[] {f[1], -1});
        }
        for (int[] f : secondList) {
            sweep.add(new int[] {f[0], 1});
            sweep.add(new int[] {f[1], -1});
        }
        Collections.sort(sweep, (a, b) -> (a[0] == b[0] ? b[1] - a[1] : a[0] - b[0]));
        
        List<int[]> list = new ArrayList<>();
        int count = 0;
        int start = 0, end = 0;
        for (int[] l : sweep) {
            count += l[1];
            if (count == 2 && l[1] == 1) {
                start = l[0];
            } else if (count == 1 && l[1] == -1) {
                end = l[0];
                list.add(new int[] {start, end});
            }
        }
        return list.toArray(new int[list.size()][]);
    }
}
